File: algorithms/sandbox/insertionSort.js
function insertionSort(arr) {
  for (let i = 1; i < arr.length; i++) {
    const currentValue = arr[i];
    let j = i - 1;

    for (j; j >= 0 && currentValue < arr[j]; j--) {
      arr[j + 1] = arr[j];
    }
    arr[j + 1] = currentValue;
  }
  return arr;
}

/* ----------
  NAIVE SOLUTION
---------- */
// function insertionSortNaive(arr) {
//   for (let head = 1; head < arr.length; head++) {
//     const currentValue = arr[head];
//     let tail = head - 1;

//     if (currentValue < arr[tail]) {
//       while (tail >= 0 && currentValue < arr[tail]) tail--;
//       arr.splice(head, 1);
//       arr.splice(tail + 1, 0, currentValue);
//     }
//   }
//   return arr;
// }


/* ----------
  TESTING
---------- */
console.log('Creating Arrays');
let start = Date.now();
const arr = new Array(100000).fill(0).map(() => Math.floor(Math.random() * 100));
// const arr2 = [...arr];
console.log(`  - ${Date.now() - start} ms`);

console.log('Running Optimized Insertion Sort');
start = Date.now();
insertionSort(arr);
console.log(`  - ${Date.now() - start} ms`);

// console.log('Running Naive Insertion Sort');
// start = Date.now();
// insertionSortNaive(arr2);
// console.log(`  - ${Date.now() - start} ms`);

File: algorithms/sandbox/insertionSort.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func main() {
	// arr := []int{1, 3, 0, 4, 2}
	// insertionSort(arr)
	// fmt.Println(arr)

	fmt.Println("Creating Arrays")
	start := time.Now()
	arr := randomArray(200000)
	fmt.Println("  - ", time.Since(start))

	fmt.Println("Running Insertion Sort")
	start = time.Now()
	insertionSort(arr)
	fmt.Println("  - ", time.Since(start))
}

func insertionSort(arr []int) []int {
	for i := 1; i < len(arr); i++ {
		currVal := arr[i]
		j := i - 1

		for j >= 0 && currVal < arr[j] {
			arr[j+1] = arr[j]
			j--
		}
		arr[j+1] = currVal
	}
	return arr
}

func randomArray(len int) []int {
	a := make([]int, len)
	for i := 0; i <= len-1; i++ {
		a[i] = rand.Intn(len)
	}
	return a
}
